pkg/upgrade/upgrades/v25_1_add_jobs_tables_test.go

```diff
@@ -0,0 +1,205 @@
+			},
+		},
+	}
+
+	ctx := context.Background()
+	tc := testcluster.StartTestCluster(t, 1, clusterArgs)
+	defer tc.Stopper().Stop(ctx)
+
+	sqlDB := sqlutils.MakeSQLRunner(tc.Conns[0])
+
+	reg := tc.Server(0).JobRegistry().(*jobs.Registry)
+	resume := make(map[jobspb.JobID]chan error)
+	defer func() {
+		for _, ch := range resume {
+			close(ch)
+		}
+	}()
+
+	defer jobs.TestingRegisterConstructor(jobspb.TypeImport, func(j *jobs.Job, _ *cluster.Settings) jobs.Resumer {
+		return jobstest.FakeResumer{
+			OnResume: func(ctx context.Context) error {
+				return <-resume[j.ID()]
+			},
+		}
+	}, jobs.DisablesTenantCostControl)()
+
+	mkJob := func(id jobspb.JobID, desc string) {
+		resume[id] = make(chan error)
+		var sj *jobs.StartableJob
+		require.NoError(t, tc.Servers[0].InternalDB().(isql.DB).Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			return reg.CreateStartableJobWithTxn(ctx, &sj, id, txn,
+				jobs.Record{
+					Details:     jobspb.ImportDetails{},
+					Progress:    jobspb.ImportProgress{},
+					Description: desc,
+					Username:    username.RootUserName(),
+				})
+		}))
+		require.NoError(t, sj.Start(ctx))
+	}
+
+	getJob := func(id jobspb.JobID) *jobs.Job {
+		j, err := reg.LoadJob(ctx, id)
+		require.NoError(t, err)
+		return j
+	}
+
+	mkJob(1000, "zero")
+	resume[1000] <- nil
+
+	mkJob(1001, "one")
+	resume[1001] <- errors.New("one failed")
+
+	// Make two jobs while still on 24.3; update their progress and status.
+	mkJob(1002, "two")
+	mkJob(1003, "three")
+	require.NoError(t, getJob(1002).NoTxn().FractionProgressed(ctx, jobs.FractionUpdater(.5)))
+	require.NoError(t, getJob(1003).NoTxn().RunningStatus(ctx, "up that hill"))
+
+	// Do this again on the verison that includes status and progress tables.
+	sqlDB.Exec(t, `SET CLUSTER SETTING version = $1`, clusterversion.V25_1_AddJobsTables.Version().String())
+	mkJob(1004, "four")
+	mkJob(1005, "five")
+
+	require.NoError(t, getJob(1004).NoTxn().FractionProgressed(ctx, jobs.FractionUpdater(.5)))
+	require.NoError(t, getJob(1005).NoTxn().RunningStatus(ctx, "up that hill"))
+
+	// And again on the version that has the new columns.
+	sqlDB.Exec(t, `SET CLUSTER SETTING version = $1`, clusterversion.V25_1_JobsWritesFence.Version().String())
+
+	mkJob(1006, "six")
+	mkJob(1007, "seven")
+	mkJob(1008, "eight")
+	require.NoError(t, getJob(1006).NoTxn().FractionProgressed(ctx, jobs.FractionUpdater(.5)))
+	require.NoError(t, getJob(1006).NoTxn().RunningStatus(ctx, "up that hill"))
+	require.NoError(t, getJob(1007).NoTxn().FractionProgressed(ctx, jobs.FractionUpdater(.5)))
+	require.NoError(t, getJob(1007).NoTxn().RunningStatus(ctx, "up that hill"))
+	resume[1006] <- nil
+	resume[1008] <- errors.New("eight failed")
+
+	jobutils.WaitForJobToSucceed(t, sqlDB, 1000)
+	jobutils.WaitForJobToFail(t, sqlDB, 1001)
+	jobutils.WaitForJobToSucceed(t, sqlDB, 1006)
+	jobutils.WaitForJobToFail(t, sqlDB, 1008)
+
+	// Let's look at the jobs table before the backfill.
```

Comment:
Note to self: when you hook up SHOW JOBS, we should run historical queries at timestamps before/during the migrations.

Reply:
We can do that, but note that the type of gates we are using in all of these changes resolve to the version value _as of the txn time_, so if you run a historical your query to before V25_1_JobsBackfill was committed, SHOW won't use any of these tables.

Reply:
right, i think this would be a good test to have in your next pr since we were burned by this during the last jobs migration.